### ISSUE TITLE
fix(modeling): Eliminate temporal leakage across chronological splits, backward-fill imputation, context routing, and stacking CV

### ISSUE THREAD

**Issue by KoshiirRa** [KoshiirRa/midgley#354]
## 1. Classification & Overview
- **Type**: Confirmed Defect / Evaluation Gap
- **Priority**: Critical (P1)
- **Reviewed Commit**: `d00ed12d61c12b7caa7c52010b29d257a3b3d48d` (Verified persistent at HEAD `7a13d2890959dab11aedf53caa1206294d287cbd`)

## 2. Problem Description & Affected Behaviour
1. **Unpurged Split Boundary**: In `src/feature_engineering.py:526-585` (`prepare_chronological_splits`), the split index is `split_idx = int(len(df) * train_ratio)`. The training target `y_train` at row `split_idx - 1` is $P_{\text{split\_idx}-1+h}$, which samples prices $h-1$ steps into `test_df`. No purging or embargo window exists at the split boundary.
2. **Backward-Fill Leakage**: In `src/feature_engineering.py:477`, missing feature values are filled with `df[feature_cols] = df[feature_cols].bfill().ffill().fillna(0.0)`. `bfill()` propagates future observations backwards into past rows.
3. **Global Diagnostic Autocorrelation**: In `src/feature_engineering.py:409-410` & `483-522` (`compute_context_routing_diagnostic`), target autocorrelation $\rho_h = \text{Corr}(X_t, X_{t+h})$ is computed over the entire dataset prior to splitting, leaking test period autocorrelation structure into training feature fusion weights.
4. **Standard K-Fold in Stacking Ensemble**: In `src/models.py:366` (`build_stacking_ensemble_pipeline`), `StackingRegressor(..., cv=5)` defaults to standard scikit-learn K-Fold CV, training on future folds to predict past folds.
5. **Future Observation Training in Purged Splitter**: In `src/models.py:605-653` (`PurgedGroupTimeSeriesSplit`), `train_mask` is initialized to `np.ones(n_samples)`, allowing observations occurring after a test fold to be included in the training indices.

## 3. Practical Impact
- Out-of-time test metrics suffer from lookahead leakage, inflating reported directional accuracy and underestimating true out-of-sample forecast error.

## 4. Permanent Code References
- `src/feature_engineering.py:526-585` (`prepare_chronological_splits`)
- `src/feature_engineering.py:477` (`bfill()`)
- `src/feature_engineering.py:483-522` (`compute_context_routing_diagnostic`)
- `src/models.py:353-367` (`build_stacking_ensemble_pipeline`)
- `src/models.py:605-653` (`PurgedGroupTimeSeriesSplit`)

## 5. Minimal Reproduction / Source Trace
```python
# src/models.py:636
train_mask = np.ones(n_samples, dtype=bool) # Includes indices > test_end!
# src/feature_engineering.py:477
df[feature_cols] = df[feature_cols].bfill().ffill().fillna(0.0) # Future data backward filled
```

## 6. Proposed Remediation
- Implement strict purging ($h$-step gap) and embargoing at every train/test split boundary.
- Replace `bfill()` with forward-fill only (`ffill().fillna(0.0)`) or historical moving statistics fitted strictly on training data.
- Fit all scalers, preprocessors, and context routing diagnostics strictly on training folds.
- Implement chronological out-of-fold cross-validation for stacking ensembles (e.g., expanding window or time-series splits respecting temporal ordering).

## 7. Checkable Acceptance Criteria
- [ ] `prepare_chronological_splits` purges all training instances whose labels overlap the test period.
- [ ] No backward filling (`bfill`) is applied to time-series feature frames.
- [ ] Preprocessors, scalers, and routing diagnostics are fitted solely on training slices.
- [ ] Stacking meta-models are trained strictly on chronological out-of-fold predictions without future lookahead.
- [ ] Unit tests assert zero data leakage and verify strict time-indexing constraints.

## 8. Dependencies & Related Issues
- Related Issues: #117 (closed precursor), #170, #314
- Downstream Dependencies: #11 (Common evaluation protocol)

---
*Documented for future implementation; no fix implemented in this task.*

**Comment by KoshiirRa** [comment]
Resolved in commit f6151a14910b897ff47bb2949e063d7083d61b14:

- **Eliminated Backward Lookahead Imputation (`bfill`):** Removed backward filling on feature matrix lags in `src/feature_engineering.py`; replaced with forward filling (`ffill()`) and neutral initializations (`fillna(0.0)`).
- **Chronological Split Boundary Purging:** Enforced an $h$-step boundary purge gap (`train_slice_end = max(1, split_idx - forecast_horizon)`) in `prepare_chronological_splits()`, guaranteeing training targets $y_t = P_{t+h}$ cannot leak test prices.
- **Training-Slice Autocorrelation Routing Diagnostic:** Evaluates target autocorrelation strictly over the training slice ($k = \lfloor N \times \text{train\_ratio} \rfloor$) in `compute_context_routing_diagnostic()`.
- **Purged Stacking Cross-Validation:** Updated `build_stacking_ensemble_pipeline()` to default to `PurgedGroupTimeSeriesSplit(n_splits=5, label_horizon_steps=5, embargo_steps=5)` with contiguous test partitions, eliminating out-of-fold training contamination in meta-regressors.
- **Verification:** Unit tests added in `tests/test_temporal_leakage.py` (5/5 passing).
